ouroboros-consensus/test-consensus/Test/Consensus/HardFork/Combinator/A.hs

```diff
@@ -248,17 +261,66 @@ instance LedgerSupportsMempool BlockA where
+          let -- note that these @ei@ invocations are all at @sno@ or earlier,
+              -- so we should see no 'PastHorizonException'
+              tipEpoch          = runEI epochInfoEpoch sno
+              firstSlotTipEpoch = runEI epochInfoFirst tipEpoch
+              epochSizeTipEpoch = runEI epochInfoSize  tipEpoch
+
+              -- The ledger must report the scheduled transition to the next
+              -- era as soon as the block containing this transaction is
+              -- immutable (that is, at least @k@ blocks have come after) --
+              -- this happens elsewhere in the corresponding 'SingleEraBlock'
+              -- instance. It must not report it sooner than that because the
+              -- consensus layer requires that conversions about time (when
+              -- successful) must not be subject to rollback.
+              --
+              -- Consensus /also/ insists that as long as the transition to the
+              -- next era is not yet known (ie not yet determined by an
+              -- immutable block), there is a safe zone that extends past the
+              -- tip of the ledger in which we guarantee the next era will not
+              -- begin. This means that we must have an additional
+              -- @safeFromTipA k@ blocks /after/ reporting the transition and
+              -- /before/ the start of the next era.
+              --
+              -- Thus, we schedule the next era to begin with the first
+              -- upcoming epoch that starts /after/ we're guaranteed to see
+              -- both the aforementioned @k@ additional blocks and also a
+              -- further @safeFromTipA k@ slots after the last of those.
+
+              -- the last slot that must be in the current era
+              firstPossibleLastSlotThisEra =
+                  History.addSlots (stabilityWindowA k + safeFromTipA k) sno
+              lastEpochThisEra =
+                  History.addEpochs
+                    (History.countSlots
+                       firstPossibleLastSlotThisEra firstSlotTipEpoch
+                     `div` unEpochSize epochSizeTipEpoch)
+                    tipEpoch
+                  `asTypeOf`
+                     -- an equivalent expression
+                     --
+                     -- The following would be equivalent if it couldn't fail
+                     -- with 'PastHorizonException', which it may since we may
+                     -- be inspecting a slot beyond the ledger's safe zone. In
+                     -- particular, the @ei@ the HFC provided to us is overly
+                     -- conservative for our specific purpose here. We're using
+                     -- it to decide here when this era should end. This era
+                     -- could only end sooner than that if this decision itself
+                     -- gets discarded! (TODO: double-check this claim)
+                     runEI epochInfoEpoch firstPossibleLastSlotThisEra
```

Comment:
I think the point is that we _should_ be able to do this calculation, whilst observing safe zones. I will try to change this.

Reply:
Ah, no, I am wrong. Now that we've spelled it out like this, it's almost _by definition_ that we are looking more than the safe zone ahead: indeed, we are looking _twice_ the safe zone ahead. Subtle. This means that the calculation for Byron must be careful here also.

Reply:
(Except of course Byron doesn't actually _use_ `EpochInfo` and just assumes epoch sizes are forever the same. That might actually be correct, in this instance -- but for Shelley to Goguen we'll need to be careful.)